### ISSUE TITLE
convenience Functions for KDFs

### ISSUE THREAD

**Issue by tauli** [haskell-crypto/cryptonite#147]
I'd like to see some convenience functions for the KDFs akin to the `encryptPass` and `verifyPass` from the package [`scrypt`](https://hackage.haskell.org/package/scrypt-0.5.0/docs/Crypto-Scrypt.html#g:2). 

If this is something that you would like to see but don't have the time to implement yourself, i'd be happy to contribute a pull request.

**Comment by tekul** [comment]
This would certainly make sense for argon2, since it has a standard format for the [encoded hash and parameters](https://github.com/P-H-C/phc-winner-argon2). The BCrypt code already has high-level functions. For scrypt it's a bit trickier, as there's no standard format that I'm aware of for encoding the hash. The Haskell `scrypt` library has invented its own, but it would probably make more sense to be compatibile with [`passlib`](http://pythonhosted.org/passlib/lib/passlib.hash.scrypt.html) and the [PHC format](https://github.com/P-H-C/phc-string-format/blob/master/phc-sf-spec.md).

**Comment by tauli** [comment]
I wasn't aware that the PHC specified a format for hash storage. I agree that that should be used. No need to define something new :-)

So i think we should add some `encryptStorable` function to `HKDF`, `PBKDF2` and `Scrypt`. I'd also think it would be neat to have a single `verify` function that sits in `Crypto.KDF` which detects the algorithm of the hash and just does the right thing. Even though BCrypt uses a slightly different format to PHC, it's easy enough to distinguish.

**Comment by tekul** [comment]
I'm not sure about `encryptStorable`. It's not really encryption. I think I would just use `hashPassword/verifyPassword` since that is the most common use case and people will immediately understand what is going on.

**Comment by tauli** [comment]
Agreed. I think I'll just implement this as soon as i find the time and send a pull request when I'm done. Probably much more productive to talk about actual code at this point.

**Comment by tekul** [comment]
Note that the format described in [that repo is wrong](https://github.com/P-H-C/phc-winner-argon2/issues/199). There's also a 'v' parameter in the string which is the argon2 version. Also rather confusingly, the version is entered as a base-16 parameter for the `argon2` command, but is shown as base-10 in the output string.

**Comment by MaxGabriel** [comment]
@tauli Did you make any progress on this?

I looked into the argon2 side, it looks like https://github.com/P-H-C/phc-winner-argon2 recommends using `argon2<v>_hash_encoded` to generate the base64 encoded version that includes parameters, but I didn't find it searching the Cryptonite codebase. Is the cryptonite C code based on an older Argon2 C library or something?

**Comment by tekul** [comment]
@MaxGabriel I wrote [some code to do this](https://github.com/ThreeMinuteLearning/my3ml/blob/d5afc4ede367222cc5c0bfb3b5f0e3ad3a608a8f/api/Password.hs) which you can use in the meantime if you wish. I will move it to cryptonite at some point when I've tidied it up and added some docs.
